### PR TITLE
feat: Support No Type

### LINKED ISSUE #168: Schema containing properties without a type cause an error
Apparently, type is not required in json schema, but target-postgres throws the error: `RuntimeError: Neither type nor anyOf are present. Unable to determine type.`

I'm not sure how common this is, or what the target should do in that situation (cast to string?), but I believe if type is not present in the schema, it's valid for any type. I ran into this using [tap-salesforce](https://github.com/meltanolabs/tap-salesforce) because salesforce has a few system/internal objects that have [fields with type "anyType"](https://github.com/MeltanoLabs/tap-salesforce/blob/d802f68929d9621a72a9639603b422ea90490520/tap_salesforce/salesforce/__init__.py#L174).

See `OldValue` and `NewValue` in the schema below.

```
{"type": "SCHEMA", "stream": "Tenant__History", "schema": {"type": "object", "additionalProperties": false, "properties": {"Id": {"type": "string"}, "IsDeleted": {"type": ["null", "boolean"]}, "ParentId": {"type": ["null", "string"]}, "CreatedById": {"type": ["null", "string"]}, "CreatedDate": {"anyOf": [{"type": "string", "format": "date-time"}, {"type": ["string", "null"]}]}, "Field": {"type": ["null", "string"]}, "DataType": {"type": ["null", "string"]}, "OldValue": {}, "NewValue": {}}}, "key_properties": ["Id"], "bookmark_properties": ["CreatedDate"]}

Traceback (most recent call last):
  File "/home/azureuser/datawarehouse/meltano/.meltano/loaders/target-postgres/venv/bin/target-postgres", line 8, in <module>
    sys.exit(TargetPostgres.cli())
  File "/home/azureuser/datawarehouse/meltano/.meltano/loaders/target-postgres/venv/lib/python3.8/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
  File "/home/azureuser/datawarehouse/meltano/.meltano/loaders/target-postgres/venv/lib/python3.8/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)     
  File "/home/azureuser/datawarehouse/meltano/.meltano/loaders/target-postgres/venv/lib/python3.8/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/azureuser/datawarehouse/meltano/.meltano/loaders/target-postgres/venv/lib/python3.8/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "/home/azureuser/datawarehouse/meltano/.meltano/loaders/target-postgres/venv/lib/python3.8/site-packages/singer_sdk/target_base.py", line 546, in invoke
    target.listen(file_input) 
  File "/home/azureuser/datawarehouse/meltano/.meltano/loaders/target-postgres/venv/lib/python3.8/site-packages/singer_sdk/io_base.py", line 33, in listen
    self._process_lines(file_input)
  File "/home/azureuser/datawarehouse/meltano/.meltano/loaders/target-postgres/venv/lib/python3.8/site-packages/singer_sdk/target_base.py", line 291, in _process_lines
    counter = super()._process_lines(file_input)
  File "/home/azureuser/datawarehouse/meltano/.meltano/loaders/target-postgres/venv/lib/python3.8/site-packages/singer_sdk/io_base.py", line 76, in _process_lines
    self._process_schema_message(line_dict)
  File "/home/azureuser/datawarehouse/meltano/.meltano/loaders/target-postgres/venv/lib/python3.8/site-packages/target_postgres/target.py", line 349, in _process_schema_message
    super()._process_schema_message(message_dict)
  File "/home/azureuser/datawarehouse/meltano/.meltano/loaders/target-postgres/venv/lib/python3.8/site-packages/singer_sdk/target_base.py", line 399, in _process_schema_message
    _ = self.get_sink(        
  File "/home/azureuser/datawarehouse/meltano/.meltano/loaders/target-postgres/venv/lib/python3.8/site-packages/singer_sdk/target_base.py", line 167, in get_sink
    return self.add_sink(stream_name, schema, key_properties)
  File "/home/azureuser/datawarehouse/meltano/.meltano/loaders/target-postgres/venv/lib/python3.8/site-packages/singer_sdk/target_base.py", line 248, in add_sink
    sink.setup()              
  File "/home/azureuser/datawarehouse/meltano/.meltano/loaders/target-postgres/venv/lib/python3.8/site-packages/target_postgres/sinks.py", line 50, in setup
    self.connector.prepare_table(
  File "/home/azureuser/datawarehouse/meltano/.meltano/loaders/target-postgres/venv/lib/python3.8/site-packages/target_postgres/connector.py", line 119, in prepare_table
    sql_type=self.to_sql_type(property_def),
  File "/home/azureuser/datawarehouse/meltano/.meltano/loaders/target-postgres/venv/lib/python3.8/site-packages/target_postgres/connector.py", line 241, in to_sql_type
    raise RuntimeError(msg)   
RuntimeError: Neither type nor anyOf are present. Unable to determine type.
```

### PR DESCRIPTION
**Pull request by andyoneal**
Uses string by default when a property in the schema does not include any type info.

fixes #168

### CODE REVIEW

File: target_postgres/tests/data_files/test_no_type.singer [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+{"type": "RECORD", "stream": "test_no_type", "record": {"Id": "0122O0001D0D08MQKX", "IsDeleted": false, "ParentId": "a301O000001CQ010AB", "CreatedById": "00501000000AaNajLAX", "CreatedDate": "2023-08-17T11:15:33.000000Z", "Field": "Used__c", "DataType": "Double", "OldValue": 11.0, "NewValue": 8.0}, "version": 1692831045028, "time_extracted": "2023-08-23T22:50:45.047299Z"}
+{"type": "RECORD", "stream": "test_no_type", "record": {"Id": "0122O0001D0D0M3QKN", "IsDeleted": false, "ParentId": "a301O000000Sx010AB", "CreatedById": "00501000000AaNajLAX", "CreatedDate": "2023-08-18T11:15:37.000000Z", "Field": "Main_Status__c", "DataType": "DynamicEnum", "OldValue": "Active", "NewValue": "Inactive"}, "version": 1692831045028, "time_extracted": "2023-08-23T22:50:45.047299Z"}
+{"type": "RECORD", "stream": "test_no_type", "record": {"Id": "0122P000001SXX0YXX", "IsDeleted": false, "ParentId": "a301O000000T601QAB", "CreatedById": "0056O00000XAxPxxLAX", "CreatedDate": "2023-08-23T14:00:56.000000Z", "Field": "Is_Bool__c", "DataType": "Boolean", "OldValue": false, "NewValue": true}, "version": 1692831045028, "time_extracted": "2023-08-23T22:50:45.047299Z"}
```
Comment (edgarrmondragon) [inline target_postgres/tests/data_files/test_no_type.singer:12]: I think it's worth adding an `object` test case here

```suggestion
{"type": "RECORD", "stream": "test_no_type", "record": {"Id": "0122P000001SXX0YXX", "IsDeleted": false, "ParentId": "a301O000000T601QAB", "CreatedById": "0056O00000XAxPxxLAX", "CreatedDate": "2023-08-23T14:00:56.000000Z", "Field": "Is_Bool__c", "DataType": "Boolean", "OldValue": false, "NewValue": true}, "version": 1692831045028, "time_extracted": "2023-08-23T22:50:45.047299Z"}
{"type": "RECORD", "stream": "test_no_type", "record": {"Id": "0122P000001SXX0YXY", "IsDeleted": false, "ParentId": "a301O000000T601QAB", "CreatedById": "0056O00000XAxPxxLAX", "CreatedDate": "2023-08-23T14:00:56.000000Z", "Field": "Is_Bool__c", "DataType": "Object", "OldValue": {"x": 1}, "NewValue": {"x": 2}}, "version": 1692831045028, "time_extracted": "2023-08-23T22:50:45.047299Z"
```

**Review by edgarrmondragon — CHANGES_REQUESTED** [review]
Let's check how lists and dicts are serialized

**Comment by andyoneal** [comment]
Good call. Lists seem to work ok, but the object test hit this error `sqlalchemy.exc.ProgrammingError: (psycopg2.ProgrammingError) can't adapt type 'dict'`

**Comment by edgarrmondragon** [comment]
> Good call. Lists seem to work ok, but the object test hit this error `sqlalchemy.exc.ProgrammingError: (psycopg2.ProgrammingError) can't adapt type 'dict'`

Yeah, I think for that to work we need to make column be of type `jsonb`. @visch any suggestions on what schema type will allow us to do that?

**Comment by visch** [comment]
> > Good call. Lists seem to work ok, but the object test hit this error `sqlalchemy.exc.ProgrammingError: (psycopg2.ProgrammingError) can't adapt type 'dict'`
> 
> Yeah, I think for that to work we need to make column be of type `jsonb`. @visch any suggestions on what schema type will allow us to do that?

We could just make the schema be an `object` but we'll have the reverse problem for strings, etc.  We need to convert the data to a string regardless of it's type (or to a jsonb object, but String probably makes the most sense here)

I see two options to move forward here (probably more are available)
1.  Add a transform function https://github.com/MeltanoLabs/target-postgres/blob/main/target_postgres/sinks.py#L149-L150 and in the appendOnly section. Detect that this field needs to be checked by adding some metadata to the jsonschema here https://github.com/MeltanoLabs/target-postgres/pull/183/files#diff-d9dd2207eaf6595663ebc6779561ea07947567ea8ecd47c504915df8f078154eR244

2. Use a custom SQLAlchemy type like this https://stackoverflow.com/questions/28143557/sqlalchemy-convert-column-value-back-and-forth-between-internal-and-database-fo for our field and transform the data there. 

Something like
```python
from sqlalchemy import TypeDecorator, String
import json

class JSONEncodedDict(TypeDecorator):
    impl = String

    def process_bind_param(self, value, dialect):
        if value is not None and isinstance(value, dict):
            value = json.dumps(value)
        return value

    def process_result_value(self, value, dialect):
        if value is not None:
            value = json.loads(value)
        return value

class YourModel(Base):
    # ...
    your_column = Column(JSONEncodedDict())
 ```

**Review by visch — CHANGES_REQUESTED** [review]
Object/Lists! :D Thanks for the PR

**Comment by andyoneal** [comment]
@visch got it to _technically_ work before I saw your suggestion here. the way I did it is hideous, so I was glad to see ideas for something more sensible. I'll likely redo this.

**Comment by visch** [comment]
Nice @andyoneal , there is something to be said about converting the data types to work if the schema thinks the data should be a string generally. 

@andyoneal I'm in the phone so it's hard for me to tell right this second but do we also have lists handled?

**Comment by andyoneal** [comment]
@visch yes, lists work too. although I am treating them the same as objects and deserializing them as json, so you end up with square brackets in the string instead of curly brackets. That's a little harder to cast from string to array in postgres.

another possibility is the way Stitch does it with their closed source target for Postgres. It seems like they're guessing the type for each record and splitting the column into correctly typed. So instead of a single string column, I've seen up to five typed variants in my data:
* ColumnName__st: text
* ColumnName__do: double precision
* ColumnName__de: numeric
* ColumnName__bo: boolean
* ColumnName__bi: bigint

We could extend that to include objects and arrays, but that could get out of hand quickly (especially with flattening turned on!) and is probably more complexity than this kind of edge case is worth.

What do you think? @visch @edgarrmondragon

**Comment by andyoneal** [comment]
Went with the sql alchemy custom type approach

File: target_postgres/connector.py [hunk 2]
```diff
@@ -237,9 +239,11 @@ def to_sql_type(jsonschema_type: dict) -> sqlalchemy.types.TypeEngine:
...
                 json_type_array.append(entry)
         else:
-            msg = "Neither type nor anyOf are present. Unable to determine type."
-            raise RuntimeError(msg)
-
+            msg = (
+                "Neither type nor anyOf are present. Unable to determine type. "
+                "Defaulting to string."
+            )
+            json_type_array.append({"type": "string", "format": "no-type"})
```
Comment (visch) [inline target_postgres/connector.py:246]:
```suggestion
            return NOTYPE()
```

File: target_postgres/connector.py [hunk 3]
```diff
@@ -270,6 +274,8 @@ def pick_individual_type(jsonschema_type: dict):
             return ARRAY(JSONB())
         if jsonschema_type.get("format") == "date-time":
             return TIMESTAMP()
+        if jsonschema_type.get("format") == "no-type":
+            return NOTYPE()
```
Comment (visch) [inline target_postgres/connector.py:278]:
```suggestion
```

**Review by visch — CHANGES_REQUESTED** [review]
Looks really good, slight tweak!

**Comment by andyoneal** [comment]
that's way simpler. thanks.

**Review by edgarrmondragon — APPROVED** [review]

**Comment by edgarrmondragon** [comment]
This LGTM. Thanks @andyoneal!

**Review by visch — APPROVED** [review]

**Pull request merged**